base/strings/search.jl

```diff
@@ -166,35 +188,40 @@ function _search_bloom_mask(c)
 end
 
 _nthbyte(s::String, i) = codeunit(s, i)
-_nthbyte(a::Union{AbstractVector{UInt8},AbstractVector{Int8}}, i) = a[i]
+_nthbyte(a::Union{Vector{UInt8},Vector{Int8}}, i) = a[i]
+_nthbyte(t::AbstractVector, index) = t[firstindex(t) + (index-1)]
 
 function _searchindex(s::String, t::String, i::Integer)
     # Check for fast case of a single byte
     lastindex(t) == 1 && return something(findnext(isequal(t[1]), s, i), 0)
     _searchindex(unsafe_wrap(Vector{UInt8},s), unsafe_wrap(Vector{UInt8},t), i)
 end
 
-function _searchindex(s::ByteArray, t::ByteArray, i::Integer)
-    n = sizeof(t)
-    m = sizeof(s)
+function _searchindex(s::AbstractVector{<:Union{Int8,UInt8}},
+                      t::AbstractVector{<:Union{Int8,UInt8}},
+                      i::Integer) where T <:Union{Int8,UInt8}
+    n = length(t)
+    m = length(s)
+    f_s = firstindex(s)
+    i < f_s && throw(BoundsError(s, i))
 
     if n == 0
-        return 1 <= i <= m+1 ? max(1, i) : 0
+        return f_s <= i <= m+1 ? max(f_s, i) : 0
     elseif m == 0
         return 0
     elseif n == 1
         return something(findnext(isequal(_nthbyte(t,1)), s, i), 0)
     end
 
     w = m - n
-    if w < 0 || i - 1 > w
+    if w < 0 || i - f_s > w
         return 0
     end
 
     bloom_mask = UInt64(0)
-    skip = n - 1
+    skip = n - f_s
     tlast = _nthbyte(t,n)
-    for j in 1:n
+    for j in eachindex(t)
```

Comment:
This isn't enough, because the `n - j - 1` computation below (and similar, like `if j == n-1`) assumes 1-based indexing.

You'll need to check all of the index calculations and fix them to use firstindex and lastindex.   Alternatively, shift the indexing calculations in nthbyte as I suggested in another comment.

Reply:
I have changed the `_nthbyte` definition thus only need to consider offset at the start and when returning the result

Reply:
In that case you should loop for `j = 1:n`